### ISSUE TITLE
consider pulling 'access' into the type of a pointer/reference

### ISSUE THREAD

**Issue by dneto0** [gpuweb/gpuweb#1604]
From discussion at https://github.com/gpuweb/gpuweb/pull/1569#issuecomment-815092147

Instead of 

      [[block]] struct B { count:  i32 };
      [[group(0), binding(0)]] var<storage> mybuf :  [[access(read)]] B;

      fn main() {
          let p: ptr<storage,i32> = &(B.count); 
          let count_copy = *p; // valid
          *p = 12;   // not valid by tracing back to the variable.
      }

Do this?

      [[block]] struct B { count:  i32 };
      [[group(0), binding(0)]] var<storage,read> mybuf : B;

      fn main() {
          let p: ptr<storage,read,i32> = &(B.count); 
          let count_copy = *p; // valid by type checking that the read is from a read ref.
          *p = 12;   // not valid, by type check because the LHS has type ref<storage,read,i32>
      }

One consideration is what happens at function boundaries.

In the original case, I would argue that the 'read' attribute should be required on the declaration of the formal parameter.  (Note: In Vulkan SPIR-V passing pointers to storage buffers is only permitted with an extension that turns on VariablePointers or VariablePointersStorageBuffer.)  There _might_ be some accommodation where a pointer to read_write buffer might be passable into a read buffer pointer formal parameter. That would be TBD.

In the second case, the default rule is no accommodation like this, and no automatic conversion of a read-write pointer to a read-only pointer.  Alternately, add a constructor (conversion) builtin that has the effect of changing the type.

cc: @kvark

**Comment by dneto0** [comment]
Another question is whether a default should be provided, for storage buffers. ('storage' storage class woohoo).
I have argued that read_write is a good default, because uniform-buffers are inherently read-only and often(?) a better choice for read-only data.

**Comment by kvark** [comment]
Having read-write being the default for pointers/references in general is totally reasonable. This isn't about storage buffers in particular.

**Comment by Kangz** [comment]
nit: Maybe have the pointer syntax be `ptr<type[, storage_class[, access]]>` so that latter type arguments can be left out without changing the order of arguments.

**Comment by alan-baker** [comment]
The default access qualifier should be storage class based. Read for uniform (and handle which is never mentioned) and read write for everything else.

We should also move the access qualifier on storage textures into the type and remove the attribute altogether. Storage textures would only allow read or write and needn't have a default. Example,
```
var t : storage_texture_2d<r32uint, read>;
```

**Comment by kvark** [comment]
@dneto0 I didn't realize that this issue takes the idea we discussed at the office hours (about adding access qualifier parameter to `ref` and `ptr`), and *mixes* it with another thing - about how vars are declared.
That other thing was discussed extensively in the past (e.g. in #1159), and I don't think it's necessary here. Since we aren't spelling the `ref` type anywhere, we can easily say that `var<storage>` at global scope has a type of `ref<T, ReadWrite>` or `ref<T, Read>` depending on the `[[access()]]` qualifier.

**Comment by ben-clayton** [comment]
> we can easily say that `var<storage>` at global scope has a type of `ref<T, ReadWrite>` or `ref<T, Read>` depending on the `[[access()]]` qualifier.

I think the goal of this is to entirely remove the `[[access()]]` qualifier, which I'm very keen to do. The more we've had to deal with `[[access()]]` as a decorator on the type, the more I feel it's in the wrong place.

The [spec currently states](https://gpuweb.github.io/gpuweb/wgsl/#attribute-access):

> Must only be applied to a type used as a store type for a variable in the storage storage class or a variable of storage texture type.

Okay, so lets start with the pattern repeatedly used in the spec:

```rust
[[block]] struct S { /* ... */ };

var<storage> s : [[access(read)]] S;
```

But what about aliases?

```rust
[[block]] struct S { /* ... */ };

type A1 = S;
var<storage> s1 : [[access(read)]] A1; // (1) Legal?

type A2 = [[access(read)]] S;  // (2) Also legal?
var<storage> s2 : A2;

type A3 = [[access(write)]] A2;  // (3) Conflicting access qualifiers, illegal?

[[block]] struct X {
   a : A2; // (4) Legal? A2 is a "type used as a store type for a variable in the storage storage class"
};

var<storage> s3 : [[access(write)]] X; // (4) again conflicting access qualifiers between s3 and s3.a
```

I assume that we *don't* want to allow anything but the first `var<storage> s : [[access(read)]] S;` case.
If that's right, it would be a whole lot simpler if the access was on the `var`, entirely ruling out any confusion w.r.t aliases, structure members, etc.

**Comment by kvark** [comment]
@ben-clayton all I'm saying is that this proposal *unnecessarily* mixes two different ideas: one that was called out, and another that just happened to be something that David was thinking about. Let's discuss them separately.

We had plenty of `[[access()]]` discussions, and I'd need to come back in time and recall all the arguments that we found back then, before considering to look at it again.

**Comment by dneto0** [comment]
> nit: Maybe have the pointer syntax be ptr<type[, storage_class[, access]]> so that latter type arguments can be left out without changing the order of arguments.

Yeah, I meant to do it that way(!)

**Comment by dneto0** [comment]
We should remove the 'access' attribute entirely. It's in the wrong place, and introduces confusions.  #1159 landed in a bad place. I was against it but got tired of arguing.

**Comment by kvark** [comment]
@dneto0 Glad to see you recharged!
How do you feel about going in 2 steps: first seeing if it makes sense to have the access as a generic parameter on pointers/references, and then we can follow-up with a discussion about having the access to be a part of the type for buffers and textures? My feeling is that the latter is a bigger discussion, and mixing them together will make it more difficult to reach consensus.

**Comment by dneto0** [comment]
> How do you feel about going in 2 steps: first seeing if it makes sense to have the access as a generic parameter on pointers/references, and then we can follow-up with a discussion about having the access to be a part of the type for buffers and textures? My feeling is that the latter is a bigger discussion, and mixing them together will make it more difficult to reach consensus.

Sounds ok to me.  It's a big sandwich and we can eat it in two or more bites.

**Comment by kvark** [comment]
Relevant discussions in the past are #1159 and #1182. We'll need to review them in order to make progress on the second bite.

**Comment by dneto0** [comment]
I've made a [slide deck](https://docs.google.com/presentation/d/1-XIH9gEjnqQJRSNnhicso7VBRd_ZuZYGwfmc4HKzM-s/edit?usp=sharing) to describe the proposal, with the reasoning behind it.

The last slide presents the proposal summary:

* Buffer: access qualifier beside storage class, with defaulting

                    var<storage,read> robuf: Params;
                    var<storage> positions: Coords;  			// read-write
                    var<storage,read_write> positions2: Coords; 	// read-write

* Storage texture: type template parameter, beside texel format

		var t1 : texture_storage_2d<rgba16float,read>; 
                  var t2 : texture_storage_2d<rgba16float,write>; 

* Ptr, ref usage inherits from the associated var
* Defaulting
    * Read-only storage → read
    * Read-write storage → read_write
* Delete ‘access’ as an attribute)

**Comment by kvark** [comment]
Thank you for dedicating your time to this presentation!

We had 2 things mixed in this issue: point access and buffer/texture type access, and now we are adding 3rd thing into the mix: the default access qualifier. Do we have to talk about all of them at once?

Read-only storage buffers are faster in WebGPU, generally speaking (unlike other APIs), because we know that, for example, we don't need any write-to-write barriers between dispatches on read-only buffers. It's also more permissive, since using it in a render pass still means users can use the parts of the same buffers for vertex data, indexes, uniforms, etc. With this in mind, I'm not seeing how defaulting to read-write is going to help. Is your motivation driven by the need to default to *something*, or you really think that read-write is a great default to offer?

btw, the "Storage buffer vs. uniform buffer" should also have a point for "use unbound array data", which is often where I see them used, and it is orthogonal to mutability.

> “A value’s type determines the syntax and semantics of operations that can be performed on that value.”
> So access restriction is tied to the type

The same argument can be applied to buffers, can't it? Assignment is also an operation.
At the end of the day, it's a matter of terminology. @dneto0 you keep repeating the arguments in your own terms, hence the arguments make sense.
To me, the real argument here is that "handle" is an implicit storage class, while "storage" is explicit. Therefore, opting for `var<read> t: texture_xx` would just be awkward.

**Comment by dneto0** [comment]
> Read-only storage buffers are faster in WebGPU

Yes, those are all great reasons to make entire buffers read-only.   To pile on another one: the implementation can turn off mechanisms for snooping writes, since no writes will be performed.

> The same argument can be applied to buffers, can't it? Assignment is also an operation.

Assignment is done through a reference type.  The reference carries the read-only-ness.  I'm making the distinction between the value-stored-in-the-box and the-reference-to-the-box. 

> At the end of the day, it's a matter of terminology. 

The definitions matter, and are meant to highlight important and meaningful distinctions.

(Aside: my vector calculus prof said "The definitions are the hard part, once you have the definitions, the rest is easy."  It's so true.)

> @dneto0 you keep repeating the arguments in your own terms, hence the arguments make sense.

So my arguments make sense.  :-)

But seriously, I assert that alternative formulations have hidden complexity or ambiguity.

> To me, the real argument here is that "handle" is an implicit storage class, while "storage" is explicit. Therefore, opting for `var<read> t: texture_xx` would just be awkward.

I don't think so.  I think the reason textures and samplers are different enough is that you are doing an indirection through a handle; that's the "right" way of thinking about it because in real life, on the hardware there is actually a mechanism mediating your access to the underlying texels. That's why you have to go through builtin functions in order to access the texels.  

You end up with multiple valid options, conceptually about where to attach the read-only-ness.  The slides argues that attaching it to the (implied) handle object is confusing because in the write-only case it makes you think that you can update the handle itself, which you can't do.  That's why you should not want to write `var<read> t: texture_xx`.  Also, what does a function call using such a thing look like?  I think it would look like this:

        var <read> t: texture_2d; // suggested read-only storage texture
        fn foo(tparam : texture_2d) -> vec4<f32> { // lost too much information here
               textureLoad(tparam, ...); // Why do I know this is allowed?
        }

and that looks bad to me.

So for storage textures, the access property be pushed out at least the storage-texture-as-a-whole.

**Comment by dneto0** [comment]
There was some offline discussion with some arguments and conclusions as follows:

* For non-texture variables, i.e. in particular storage buffers, it's fine to move the access property to the 'var' part of the declaration, e.g. beside the storage class when it's there.  (In agreement with slides proposal.).
* For defaulting the access property:
     * Defaulting access property enhances authorability in many cases.
     * It's fine to the default access as a function of storage class.  (It doesn't have to be as broadly "consistent" as read-only vs. read-write.  Contrary to slides proposal.)
    * Because read-write carries costs with wide implications for performance and discoverability of performance problems, the group prefers to default storage buffer to *read-only*. (Contrary to slides proposal.)
    * For storage textures, there was no desire to default the access property.

* Group prefers to pass samplers and textures to helper functions by (handle) value, not by pointer.
     * This makes it nicer to embed the access property into the type itself. (In agreement with slides proposal.).
     * In future we may have large arrays of textures or samplers, including a bindless approach.   At that time we can revisit whether we also want to add passing by pointer/reference.

* Discussed whether to allow passing a read-write pointer to a function argument that is a read-only pointer.
    * Discussed that this could be done either implicitly (an automatic conversion), or explicitly via constraining-cast.
    * The group rejected the feature.   That may be nice in a systems language, but we decided we didn't need that much flexibility in a shader language. We can revisit this in the future.

* Discussed the challenge of translating Vulkan SPIR-V to WGSL
    * @dneto's strategy (in the Tint SPIR-V reader)  is to analyze texture usage, propagating usages and constraints interprocedurally. If there are conflicts, then reject the shader as untranslatable. This is already something developers porting their applications to WebGPU have to contend with, where WebGPU deliberately subsets the functionality available in other APIs.
    * The alternative choice is to do function specialization, but that might blow up combinatorially when a helper function has many texture operands.

**Comment by dneto0** [comment]
If I understood correctly, the amended proposal is:

* Buffer: access qualifier beside storage class, with defaulting

                  var<storage,read> robuf: Params;
                  var<storage> positions: Coords;  			// read-only is the default
                  var<storage,read_write> positions2: Coords; 	// read-write

* Storage texture: type template parameter, beside texel format.  It is not defaulted.

                  var t1 : texture_storage_2d<rgba16float,read>; 
                  var t2 : texture_storage_2d<rgba16float,write>; 

 
* Defaulting, by storage class
    * `uniform`, `handle` → read    
          * uniform buffers are read-only
          * you can't write a handle (texture or sampler handle); the handles themselves are read-only
    * `storage`  ->  read.    This is the fast path
    * `function`, `private`, `workgroup`  -> read_write.    Assume there is no cost to having them read-write.

* Ptr, ref access property
    * An instantiation of a ptr, ref inherits the access property from underlying ptr, ref, or ultimately the originating variable.
    * The defaulting rules apply, inherited from the defaulting rule for the underlying storage class.

* There is no way to convert a pointer (respectively, ref) with one access property into a pointer (respectively ref) with a different access property.
    
* Delete ‘access’ as an attribute

**Comment by kvark** [comment]
As mentioned on the office hours, I think it's important to have a solid SPIR-V to WGSL story before we can proceed with this approach.

Imagine a function in SPIR-V that has N pointers in the arguments, and it writes to none of them (i.e. only does `OpLoad`). Since we can trace all the "texture" type values back to the bindings (at least right now), we can know, for each invocation, whether the relevant argument declaration in WGSL needs to be "read-write" or "read-only". So we may need to have 2^N duplicates of this function in the produced WGSL, in the worst case, given the combination of the arguments.

@dneto0 what is your suggested path here? Are you saying that a translator would *never* duplicate a function? It's fairly easy to construct a SPIR-V that would make sense and that would be rejected by this.

**Comment by alan-baker** [comment]
If we allow a `read_write` pointer to match a `read`, `write` or `read_write` function parameter then translation from SPIR-V is not particularly painful. It also has utility in WGSL for developers writing helper functions. The analysis is local to the function (assuming the SPIR-V functions are processed topologically).

**Comment by dneto0** [comment]
> Imagine a function in SPIR-V

I was asserting the interprocedural analysis is only done to determine usage of a texture: e.g. whether it's a sampled texture, read-only storage texture, write-only storage texture.

> Imagine a function in SPIR-V that has N pointers in the arguments, and it writes to none of them (i.e. only does OpLoad). Since we can trace all the "texture" type values back to the bindings (at least right now)...

I'm confused whether you're talking about storage textures, or not.

If you're talking about non-textures, then:
- for storage buffer the NonWritable attribute tells you whether it's read-only. (In SPIR-V the the non-writable attribute is on the variable or on the struct member type. ) (yeah, fun... reasons). 
- otherwise, use the storage class default as above.
In other words, this is a consequence of the fact that lower level languages like LLVM IR and SPIR-V don't have 'const'-ness as a concept.  (An SSA value is constant during its lifetime/scope;  and memory is either read-write or read-only.  You can't have a pointer-to-const-foo when the underlying foo is read-write.)

If you're talking about storage textures then, do interprocedural analysis:
- if the same bound texture variable is used for both OpImageRead and OpImageWrite, then error out: we would translate that to a read-write texture but that's not supported by WGSL.
- otherwise, are you talking a different problem?  If so can you write it out more explicitly?

**Comment by kvark** [comment]
The "texture" mention there was incorrect. I started writing this for textures, changed to just pointers in the middle, and missed this one place, sorry about the confusion!

Here is the example of a function (imagine this is SPIR-V):
```wgsl
fn foo(a1: ptr<i32>, a2: ptr<i32>, a3: ptr<i32>) {..}
```
If we go with the approach of "access has to match the variable", then the generated WGSL code may need to have different versions of this function (some with `ptr<i32, read>`, others with `ptr<i32, read_write>`), based on the concrete arguments it's called with (buffers).

If we go with the approach that allows "read_write" to be converted to "read", then it looks like SPIRV -> WGSL path would be straightforward (echoing what @alan-baker said).

**Comment by dneto0** [comment]
Ah, that makes more sense.

Good question.

- In baseline Vulkan SPIR-V, you can't pass pointer-to-storage-buffer into a helper function. So the case doesn't arise
- If you add VariablePointers or VariablePointersStorageBuffer capability, then you *can* pass pointer-to-storage-buffer into a helper function
       - The main compiler (that I know of) that emits code with those capabilities is Clspv.  Clspv does not distinguish between read-only storage buffers and read-write storage buffers
       - So the problematic case doesn't arise.
       - You might ask: why didn't users of Clspv care?  My guess that the logic goes like this:  those are users coming from OpenCL, and in OpenCL the pipeline barriers (i.e. between commands) are implicit, so there was no expectation of performance gain by making the read-only buffer distinction.

(As always, @alan-baker should feel free to correct me.)

**Comment by kvark** [comment]
>In baseline Vulkan SPIR-V, you can't pass pointer-to-storage-buffer into a helper function. So the case doesn't arise

It doesn't matter if you can't pass the pointer to the whole storage buffer. The pointers may be into parts of the storage buffer. This is well supported in SPIR-V. In fact, **all** the SPIR-V generated from GLSL passes data into functions by pointers.

**Comment by alan-baker** [comment]
> > In baseline Vulkan SPIR-V, you can't pass pointer-to-storage-buffer into a helper function. So the case doesn't arise
> 
> It doesn't matter if you can't pass the pointer to the whole storage buffer. The pointers may be into parts of the storage buffer. This is well supported in SPIR-V. In fact, **all** the SPIR-V generated from GLSL passes data into functions by pointers.

I'm not sure why you think SPIR-V allows this. The rules without variable pointers in universal validation only allow `Function`, `Private`, `UniformConstant`, `Workgroup`, and `AtomicCounter` (not in WGSL) storage class pointers as parameters.

**Comment by kvark** [comment]
This is the case I'm talking about (imagine this is SPIR-V):
```rust
fn foo(a1: ptr<storage, i32>, a2: ptr<storage, i32>, a3: ptr<storage, i32>) {
  // do some loads from a1, a2, a3
}
struct Buffer {
  data: array<i32, 5>;
};
var<storage> buffer: Buffer;
...
foo(&buffer.data[0], &buffer.data[1], &buffer.data[2]);
```
The `foo` declaration in SPIR-V doesn't care whether the pointers point to parts of read-only storage variables or not. But if we force the `access` into the pointer type, and we require that this access *matches* the owning buffer variable, then we may end up needing to generate all the permutations in WGSL:
```rust
fn foo_rrr(a1: ptr<storage, i32, read>, a2: ptr<storage, i32, read>, a3: ptr<storage, i32, read>) {..}
fn foo_wrr(a1: ptr<storage, i32, read_write>, a2: ptr<storage, i32, read>, a3: ptr<storage, i32, read>) {..}
fn foo_rwr(a1: ptr<storage, i32, read>, a2: ptr<storage, i32, read_write>, a3: ptr<storage, i32, read>) {..}
fn foo_wwr(a1: ptr<storage, i32, read_write>, a2: ptr<storage, i32, read_write>, a3: ptr<storage, i32, read>) {..}
...
```

**Comment by alan-baker** [comment]
That's invalid SPIR-V without the `VariablePointersStorageBuffer` capability declared.

**Comment by alan-baker** [comment]
Also, #1674 disallows pointer in the `storage` storage class for the same reason.

**Comment by kvark** [comment]
Interesting, thanks for clarification!
Do I understand correctly then:
  1. even if we require the pointer access to match the buffer variable access, there is 1:1 translation path from SPIR-V because we don't have variable pointers?
  2. when we *do* get variable pointers, this will be come a problem?

I thought we are going to get the variable pointer semantics at some point. Just don't want us to draw ourselves into a corner here. Or maybe, when we decide to add support for variable pointers, we can at the same time say that the read_write pointers can be coerced to "read" as a requirement?

**Comment by alan-baker** [comment]
1. Yes
2. Yes. I think the the coercion can deal with this most efficiently though.

**Comment by kvark** [comment]
Ok, sounds like a plan :)

**Comment by dneto0** [comment]
>  Or maybe, when we decide to add support for variable pointers, we can at the same time say that the read_write pointers can be coerced to "read" as a requirement?

An alternative that we've discussed in meetings, but not yet written down (that I know of), and not a proposal is:

- Add a syntax to *explicitly* coerce a ptr<SC,T,read_write> to a ptr<SC,T,read>

I know @litherum and @jimblandy cautioned against this because now it's about what the _shader_ is allowed to modify vs. what any agent anywhere is allowed to modify.  That is, allowing this would weaken the guarantees that the compiler is relying on when compiling the helper function:

- Before, the compiler knows the memory underlying the parameter will not change.
- After, the compiler knows that the shader invocation itself won't change the contents of the memory, but some other invocation or agent might.

Off the cuff @alan-baker and I argued that doing so in an unsynchronized way is a data race anyway.  This discussion is not resolved yet.

**Comment by jdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/WGSL-2021-05-11-Minutes#consider-pulling-access-into-the-type-of-a-pointerreference-1604">WGSL meeting minutes 2021-05-11</a></summary>



*   New proposal: [https://github.com/gpuweb/gpuweb/issues/1604#issuecomment-832937987](https://github.com/gpuweb/gpuweb/issues/1604#issuecomment-832937987) 
*   AB:: Seem close to consensus.  Related is passing textures/samplers by handle-value vs. pointer.
*   MM: Is it valid to take a const access to a non-const.
*   AB: Would like to be able to get a read-reference to a read-write buffer.
*   JG: Hear there are two schools of thought.
*   MM: Doing that genericity is an accidental feature. Think for MVP it’s ok to make a function not generic over const vs. non-const
*   AB: Fine with that. Not big enough usability gain.
*   ~~Consensus: access must match (ie no read_write -> read decay)~~
*   MM: Had been thinking about allowing a function to be    ptr with ‘*’ for read, write, or read_write.  That’s what we think we shouldn’t tackle now.
*   **Consensus: Must declare access qualifiers (no templating/generic functions (yet))**
*   MM: So now, can we make a ptr&lt;read> from a ptr&lt;read_write>
*   **Consensus: no read_write -> read decay for now**
*   MM: So some background: One thing to consider here is that the contents behind the read-only might be changed elsewhere.
*   JB: Probably less bad so long as we don’t call it ‘const’ :)
*   MM: Yeah, difference between “I cannot change the object” vs “The object cannot be changed”
*   AB: Proposal: Forbid (for now) passing e.g. textures by value
*   MM: Tentative agreement, but want more time to think.
*   DN: So is my proposal the direction? (modulo minor things?) (yes)


</details>
